### ISSUE TITLE
cmake configuration guidance

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#31]
Outside this medium we had an exchange where I needed you to set up downloading and inclusion of an open source yaml parser library.   It would be helpful if you could point me and anyone else reading this to an appropriate online source that provides guidance on how to do this in general.  Altrnatively or also, it would be helpful here to at least document the pieces that need to be set up.   From my attempt I know it is at least two things:

1. Changes to the file mspass/CMakeList.txt that defines the overall control of how this is done (I think)
2. To work it this framework a package needs configuration files built that are installed in mspass/cmake.   How those are used and which are required was not at all clear to me from any definitions in the existing files.   

Are there other files that need to be changed/added?

**Comment by pavlis** [comment]
Case in point:   I tried to use yaml-cpp after you set up the configuration, and I get this message when trying to compile MetadataDefinitions using a new yaml parser:
/home/pavlis/src/mspass/cxx/src/lib/utility/MetadataDefinitions.cc:5:10: fatal error: yaml-cpp/yaml.h: No such file or directory
 #include "yaml-cpp/yaml.h"
          ^~~~~~~~~~~~~~~~~
compilation terminated.

cmake did correctly download and build yaml-cpp, but something got messed up in the path set up for include and presumably lib files as well.   i.e.  can see the include file requested here under build, but some path is wrong.   I don't see how that is being set up by cmake.   Can you correct this and let me know what you needed to do to fix it?

**Comment by wangyinz** [comment]
That is strange. I tried adding the include file and it worked for me. Here is output of `git diff` from what I did:
```
--- a/cxx/src/lib/utility/CMakeLists.txt
+++ b/cxx/src/lib/utility/CMakeLists.txt
@@ -1,10 +1,11 @@
 FILE(GLOB sources_utility *.cc)
 include_directories(
   ${Boost_INCLUDE_DIRS}
+  ${YAML_CPP_INCLUDE_DIR}
   ${PROJECT_SOURCE_DIR}/include
   ${PROJECT_BINARY_DIR}/include)

 add_library(utility STATIC ${sources_utility})
-target_link_libraries(utility PRIVATE ${BLAS_LIBRARIES})
+target_link_libraries(utility PRIVATE ${BLAS_LIBRARIES} ${YAML_CPP_LIBRARIES})

 install (TARGETS utility DESTINATION lib)
diff --git a/cxx/src/lib/utility/MetadataDefinitions.cc b/cxx/src/lib/utility/MetadataDefinitions.cc
index f3987ae..976db99 100644
--- a/cxx/src/lib/utility/MetadataDefinitions.cc
+++ b/cxx/src/lib/utility/MetadataDefinitions.cc
@@ -3,6 +3,7 @@
 #include "mspass/utility/MsPASSError.h"
 #include "mspass/utility/AntelopePf.h"
 #include "mspass/utility/MetadataDefinitions.h"
+#include "yaml-cpp/yaml.h"
 namespace mspass{
 using namespace mspass;
 using std::string;
```
Did you change the `CMakeLists.txt` for `utility`?

**Comment by pavlis** [comment]
Thanks for handling that so quickly.   The example of the fix you created is good education for me on cmake.

Good news is all that works and I've managed to work through all the compile errors I had on my initial attempt at writing a yaml reader for MetadataDefinitions.   The bad news is I'm getting link errors I've not gotten from any other package we have installed by a similar mechanism.   Here is the message I'm getting when mspasspy is trying to link with the new code that uses yaml-cpp:

[ 76%] Linking CXX shared module mspasspy.cpython-36m-x86_64-linux-gnu.so
/usr/bin/ld: ../yaml-cpp/lib/libyaml-cpp.a(scanscalar.cpp.o): relocation R_X86_64_PC32 against symbol `_ZNK4YAML5RegEx14MatchUncheckedINS_16StreamCharSourceEEEiRKT_' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
python/CMakeFiles/mspasspy.dir/build.make:101: recipe for target 'python/mspasspy.cpython-36m-x86_64-linux-gnu.so' failed
make[2]: *** [python/mspasspy.cpython-36m-x86_64-linux-gnu.so] Error 1
CMakeFiles/Makefile2:88: recipe for target 'python/CMakeFiles/mspasspy.dir/all' failed
make[1]: *** [python/CMakeFiles/mspasspy.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2

The suggestion by the compiler to use -fPIC  is not the issue, I think, when I poked around the cmake files and found the link line I think is creating this error:

pavlis-home-linux:~/src/mspass/cxx/build/python/CMakeFiles/mspasspy.dir> cat link.txt
/usr/bin/c++ -fPIC -Wall -O0 -g  -shared  -o mspasspy.cpython-36m-x86_64-linux-gnu.so CMakeFiles/mspasspy.dir/mspass_wrapper.cpp.o ../src/lib/libmspass.a ../src/lib/seismic/libseismic.a ../src/lib/algorithms/libalgorithms.a ../src/lib/utility/libutility.a ../yaml-cpp/lib/libyaml-cpp.a -llapack -lblas 

-fPIC is in that command list, but the issue seems to be that all the libraries referenced are static (.a).   I think we might need to set up the configuration to create a -shared library from at least the yaml-cpp package.   Not sure though.  Any ideas you might have would be appreciated.

**Comment by wangyinz** [comment]
Well, I don't really understand this. We probably shouldn't link a dynamic library to the python module as that is much easier to break. Maybe the best way to figure this out is to push what you have to a new branch so that I can poke around.

**Comment by wangyinz** [comment]
> Outside this medium we had an exchange where I needed you to set up downloading and inclusion of an open source yaml parser library. It would be helpful if you could point me and anyone else reading this to an appropriate online source that provides guidance on how to do this in general. Altrnatively or also, it would be helpful here to at least document the pieces that need to be set up. From my attempt I know it is at least two things:
> 
> 1. Changes to the file mspass/CMakeList.txt that defines the overall control of how this is done (I think)
> 2. To work it this framework a package needs configuration files built that are installed in mspass/cmake.   How those are used and which are required was not at all clear to me from any definitions in the existing files.
> 
> Are there other files that need to be changed/added?

Forgot to answer this question. Actually, I don't have a good general reference for what I did there, since I basically followed a mixture of some other people's code. But, basically, what I did is really using the [`ExternalProject`](https://cmake.org/cmake/help/latest/module/ExternalProject.html) feature of CMake. Maybe [this](https://foonathan.net/2016/07/cmake-dependency-handling/) and [this](https://stackoverflow.com/questions/28346530/cmake-automatic-boost-download-and-build-if-not-found) are also helpful.

**Comment by wangyinz** [comment]
According to this [thread](https://stackoverflow.com/questions/19768267/relocation-r-x86-64-32s-against-linking-error), the problem could be that `yaml-cpp` is not compiled with `-fPIC`. I just quickly looked at its configuration, and it seems that is the case. I guess I can add that compile flag to it. However, the question will become what if a user already got a `yaml-cpp` built by himself on the system.

**Comment by pavlis** [comment]
First, thanks for those links.  They are good sources and it is good to preserve them here for future reference.  Better than yet another bookmark.

For the yaml problem I'll try to see if I can figure out why(if) yaml is not being compiled with -fPIC.   If I can't figure it out I'll push the branch I have with the yaml parser code.

**Comment by pavlis** [comment]
Well, I spent an hour and a half trying to figure this out and failed.   I'd like to push it back to you hoping you can solve this more quickly than I can.   Note [this](https://stackoverflow.com/questions/1340402/how-can-i-tell-with-something-like-objdump-if-an-object-file-has-been-built-wi) may be helpful, but not sure it applies to a .a file.   I'm pushing it too you because I am not clear on how to do this with the cmake configuration files and I've gotten frustrated so perhaps a new head attacking the problem will solve the problem.   

I pushed a new branch with an obvious name that you can use to verify the problem.

**Comment by wangyinz** [comment]
Turns out we just need to add `-DCMAKE_POSITION_INDEPENDENT_CODE=ON` to yaml-cpp's configure line.

**Comment by pavlis** [comment]
Impressive - glad you figure that out.   Now I can debug the yaml reader for MetadataDefinitions.

**Comment by pavlis** [comment]
I need your help on this one.  I probably could solve this, but I'm 100% sure you can solve this more efficiently than me.   You will recall you found the "dpss" open source package for C++ that you used for multitaper methods.   Figured out (well, maybe not completely - more on that below) how to get cmake to find gsl that was needed for the deconvolution code.   However, when the compilation gets to trying to compile the dpss.cc code I am getting this error:
```
[ 40%] Building CXX object src/lib/deconvolution/CMakeFiles/deconvolution.dir/dpss.cc.o
/Users/pavlis/src/mspass/cxx/src/lib/deconvolution/dpss.cc:39:21: error: use of
      undeclared identifier 'dlamch_'
    double abstol=2*dlamch_(&safe);         //safe minimum absolute tolerance
                    ^
/Users/pavlis/src/mspass/cxx/src/lib/deconvolution/dpss.cc:54:5: error: use of
      undeclared identifier 'dstebz_'
    dstebz_(&range, &order, &n, &vl, &vu, &il, &iu, &abstol, D, E, &m, &...
    ^
/Users/pavlis/src/mspass/cxx/src/lib/deconvolution/dpss.cc:56:9: error: use of
      undeclared identifier 'dstein_'
        dstein_(&n, D, E, &m, eig_val, IBLOCK, ISPLIT, eig_vec, &ldz, WO...
        ^
3 errors generated.
```
Those are lapack fortran functions that were pulled from perf.h when we were using this with antelope.   Not sure if the solution to this is some include from lapack, FC.h, or both.  You cracked the lapack/openblas months ago.  That is why I'm certain you can solve this problem much more efficiently than me.

**Comment by pavlis** [comment]
Ignore the question about gsl.  I now know that works as when I pulled the add_deconvolution branch to my ubuntu machine it built gsl successfully.

**Comment by pavlis** [comment]
Well, wrong about that, but think I can solve it.  I appear to need to find a cmake configuration file on the web somewhere.

**Comment by wangyinz** [comment]
The gsl related cmake configuration should be fixed by my recent commits. Are you having trouble from the latest commits?

**Comment by pavlis** [comment]
It appears something is wrong with the new gsl cmake setup.    I found it works fine with MacOS where it somehow found a version of gsl local.  At some point you need to teach me how it does that.  The problem was when I tried to build on my ubuntu machine.  It is downloading the gsl library and compiling it in the build directory.  After it finishes compiling it aborts with this message from cmake:
```
/usr/bin/install -c -m 644 gsl_math.h gsl_pow_int.h gsl_nan.h gsl_machine.h gsl_mode.h gsl_precision.h gsl_types.h gsl_version.h gsl_minmax.h gsl_inline.h '/home/pavlis/src/mspass/cxx/build/gsl/include/gsl'
[ 88%] No test step for 'gsl'
[100%] Completed 'gsl'
[100%] Built target gsl
CMake Error at cmake/gsl.cmake:25 (find_package):
  By not providing "Findgsl.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "gsl", but
  CMake did not find one.

  Could not find a package configuration file provided by "gsl" with any of
  the following names:

    gslConfig.cmake
    gsl-config.cmake

  Add the installation prefix of "gsl" to CMAKE_PREFIX_PATH or set "gsl_DIR"
  to a directory containing one of the above files.  If "gsl" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:51 (fetch_gsl)


-- Configuring incomplete, errors occurred!
See also "/home/pavlis/src/mspass/cxx/build/CMakeFiles/CMakeOutput.log".
```
This was the error I got earlier I thought I could solve, but I couldn't find any useful reference the web that could give me a hint about where I could get the gslConfig.cmake file what should be in it.   

Curiously I also installed gsl with apt install.  That completed successfully, but after doing so the cmake process did not find the local system installation but went through the download and compile sequence again throwing the above error.  

So, one solution is to better understanding what directories find_package search and how they are configured.   Then I could use the local copy on this machine.   I now have a workaround that makes this less critical.   I can work on my laptop and do some initial debugging of the deconvolution stuff.   I need to next build a C++ test suite.   I had something for the antelope version, but it has dependencies on the stream io library that make it impossible to port.   Let me know if you can come up with a solution for this, but don't waste your weekend.   This can wait.

**Comment by wangyinz** [comment]
What version of cmake are you using? I suspect this is an issue with old version.

**Comment by pavlis** [comment]
I am running a fairly old version 3.1.2 but that is what is current binary release from running apt.   Do I need to install a newer rev from source code?

**Comment by wangyinz** [comment]
Yeah, that is absolutely too old. I think the support of gsl comes in 3.4 or something. It is likely that the Ubuntu you have is also a relatively old version. We are using Ubuntu bionic in the docker container.

**Comment by pavlis** [comment]
This problem does not seem to be limited to the version of cmake.  I installed much newer version on my ubuntu machine.   Further, after some confusion created by a system problem that apparently went away when the did system maintenance today, I get the same error on the carbonate machine.   This may be a trivial fix, but I'm stumped.   I really need to find some kind of more comprehensive document on cmake so I'm no just poking in the dark and resorting to google searches to fix problems like this.  Anyway, here is the message I get consistently on linux ubuntu and redhat (carbonate):

100%] Built target gsl
CMake Error at cmake/gsl.cmake:25 (find_package):
  By not providing "Findgsl.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "gsl", but
  CMake did not find one.

  Could not find a package configuration file provided by "gsl" with any of
  the following names:

    gslConfig.cmake
    gsl-config.cmake

  Add the installation prefix of "gsl" to CMAKE_PREFIX_PATH or set "gsl_DIR"
  to a directory containing one of the above files.  If "gsl" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:55 (fetch_gsl)


-- Configuring incomplete, errors occurred!
See also "/N/u/pavlis/Carbonate/src/mspass/cxx/build/CMakeFiles/CMakeOutput.log".

Note in both cases gsl is improperly built.  I had gsl installed on my ubuntu machine as a standard package. On carbonate I had their "modules" system set to know about gsl, but cmake is simply not finding it.  Then after it builds it's own copy it can't seem to configure itself.   Any ideas?

**Comment by wangyinz** [comment]
OK, I found the issue and has pushed the fix. Turns out it is because I used lower case "gsl" in the find_package command while the actual script is expecting "GSL", and it seems cmake is case sensitive (which makes me wonder why it worked inside the container). Anyway, that should get this issue fixed. 

In terms of the cmake documentation, I don't think the automatic installation script that I used is recommended or included in any of their official documents. I just adopted this method based on the setup that I've seen some packages uses.

**Comment by pavlis** [comment]
Thanks so much for that fix. Explains why google didn't work for this one - simple rookie error by both of us.   
Incidentally, on carbonate system updates they completed yesterday made the added arguments to cmake no longer needed.   You do, however, need to unload default gcc and load the newest version they support.

**Comment by pavlis** [comment]
Do you have any idea how to silence this kind of bogus warning message:
```
    if(nsamp>d.size())
/home/pavlis/src/mspass/cxx/include/mspass/deconvolution/ComplexArray.h:180:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```
Any time a C++ program uses the std::vector size method in a construct like that where a regular
int is compared to the output you cget the -Wsign-compare warning above.   That construct is very common and a somewhat silly warning.  It probably should only be present for if one of my favorite compiler flag is set:  -Wpedantic.  There is an obscure English word for you, but it totally describes this particular error.  Why comparison to a signed and unsigned int should generate a warning by default is beyond me since even basic programming books contain such constructs.   

Fixing this requires some global parameter change for cmake.  I think it is probably comes from having the CMAKE_FLAGS_DEBUG argument set to -Wall, but I'm reluctant to change that because it will suppress some other warning that can be useful.

Note this seems to only be an issue in newer versions of gcc.  Older revs do not behave this way.

**Comment by wangyinz** [comment]
Yep, I think it is exactly from -Wall. We should probably just remove that from this line: https://github.com/wangyinz/mspass/blob/c6b354ebc70e8e133a6886f0f33ff34b8e68cb42/cxx/CMakeLists.txt#L13

or, we might just need to add a -Wno-sign-compare flag there.

**Comment by pavlis** [comment]
I've added the no-sign-compare flag to the add_deconvolution branch where it is most needed. We can decide when we are ready to merge if that is a good idea overall.

**Comment by pavlis** [comment]
In our call today we discussed the oddity I was seeing with boost libraries not being resolved in builds of the ccore so file I creat in Ubuntu.   Have been using LD_LIBRARY_PATH to resolve the issue.   I have some new developments on this issue, but haven't really solved it.  The good news is it would be unlikely to be an issue for someone installing mspass from source code.   

What I found was this.   
1.  When I did the install solution you gave me to ~/.local I did not have the problem.  
2.   Next build I copied the ccore so file to the appropriate place in the local install as you suggested.   Then the problem resurfaced.
3.  I have verified the CMakeList.txt file in the add_deconvolution branch that defines boost (the top level file in the cxx directory) is identical to master.   

I'm going to try the --upgrade option with pip for the local install to see if that works without inducing a complete new compilation.
